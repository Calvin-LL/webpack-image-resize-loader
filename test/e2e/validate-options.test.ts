import webpack from "webpack";

import WIRLWebpackTestCompiler from "./helpers/WIRLWebpackTestCompiler";

describe.each([4, 5] as const)("v%d validate options", (webpackVersion) => {
  const tests = {
    width: {
      success: [10],
      failure: [true, 0, 0.1],
    },
    height: {
      success: [10],
      failure: [true, 0, 0.1],
    },
    fit: {
      success: ["cover", "contain", "fill", "inside", "outside"],
      failure: [true, "0"],
    },
    position: {
      success: [
        "top",
        "right top",
        "right",
        "right bottom",
        "bottom",
        "left bottom",
        "left",
        "left top",

        "north",
        "northeast",
        "east",
        "southeast",
        "south",
        "southwest",
        "west",
        "northwest",
        "center",
        "centre",

        "entropy",
        "attention",
      ],
      failure: [true, "0"],
    },
    background: {
      success: ["#FFF", "rgb(2,4,99)", { r: 255, g: 255, b: 255 }],
      failure: [false, 9],
    },
    scale: {
      success: [0.1, 0.5, 1],
      failure: [-1, 0],
    },
    format: {
      success: ["jpeg", "png", "webp", "tiff"],
      failure: [true],
    },
    quality: {
      success: [10],
      failure: [0],
    },
    scaleUp: {
      success: [true, false],
      failure: [0],
    },
    sharpOptions: {
      success: [{ resize: { fastShrinkOnLoad: true } }],
      failure: [0],
    },
    imageminOptions: {
      success: [
        { png: null },
        { png: ["imagemin-optipng"] },
        { png: [{ name: "imagemin-optipng" }] },
        {
          png: [
            {
              name: "imagemin-optipng",
              options: {
                optimizationLevel: 1,
              },
            },
          ],
        },
      ],
      failure: [0],
    },
    fileLoader: {
      success: ["file-loader"],
      failure: [0],
    },
    fileLoaderOptionsGenerator: {
      success: [(_: any, o: any) => o],
      failure: [0],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ): void {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = new WIRLWebpackTestCompiler({ webpackVersion });

      let stats: webpack.Stats | undefined;

      try {
        stats = (
          await compiler.compile({
            loaderOptions: {
              width: 10,
              [key]: value,
            },
            throwOnError: false,
          })
        ).stats;
      } finally {
        if (type === "success") {
          expect(stats!.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const errors = stats!.compilation.errors;

          expect(errors).toHaveLength(1);
          expect(errors[0].error.message).toMatchSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
