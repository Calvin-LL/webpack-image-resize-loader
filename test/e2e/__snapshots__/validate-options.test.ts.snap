// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v4 validate options should throw an error on the "background" option with 9 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v4 validate options should throw an error on the "background" option with false value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v4 validate options should throw an error on the "fileLoader" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoader should be a string."
`;

exports[`v4 validate options should throw an error on the "fileLoaderOptionsGenerator" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoaderOptionsGenerator should be one of these:
   string | function
   Details:
    * options.fileLoaderOptionsGenerator should be a string.
    * options.fileLoaderOptionsGenerator should be an instance of function."
`;

exports[`v4 validate options should throw an error on the "fit" option with "0" value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fit should be one of these:
   \\"cover\\" | \\"contain\\" | \\"fill\\" | \\"inside\\" | \\"outside\\""
`;

exports[`v4 validate options should throw an error on the "fit" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fit should be one of these:
   \\"cover\\" | \\"contain\\" | \\"fill\\" | \\"inside\\" | \\"outside\\""
`;

exports[`v4 validate options should throw an error on the "format" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.format should be one of these:
   \\"jpeg\\" | \\"png\\" | \\"webp\\" | \\"avif\\" | \\"tiff\\""
`;

exports[`v4 validate options should throw an error on the "height" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be >= 1."
`;

exports[`v4 validate options should throw an error on the "height" option with 0.1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be >= 1."
`;

exports[`v4 validate options should throw an error on the "height" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be a number (should be >= 1)."
`;

exports[`v4 validate options should throw an error on the "position" option with "0" value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.position should be one of these:
   \\"top\\" | \\"right top\\" | \\"right\\" | \\"right bottom\\" | \\"bottom\\" | \\"left bottom\\" | \\"left\\" | \\"left top\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\" | \\"center\\" | \\"centre\\" | \\"entropy\\" | \\"attention\\""
`;

exports[`v4 validate options should throw an error on the "position" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.position should be one of these:
   \\"top\\" | \\"right top\\" | \\"right\\" | \\"right bottom\\" | \\"bottom\\" | \\"left bottom\\" | \\"left\\" | \\"left top\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\" | \\"center\\" | \\"centre\\" | \\"entropy\\" | \\"attention\\""
`;

exports[`v4 validate options should throw an error on the "quality" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.quality should be >= 1."
`;

exports[`v4 validate options should throw an error on the "scale" option with -1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scale should be > 0."
`;

exports[`v4 validate options should throw an error on the "scale" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scale should be > 0."
`;

exports[`v4 validate options should throw an error on the "scaleUp" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scaleUp should be a boolean."
`;

exports[`v4 validate options should throw an error on the "sharpOptions" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.sharpOptions should be an object:
   object { resize?, png?, jpeg?, webp?, tiff?, … }"
`;

exports[`v4 validate options should throw an error on the "width" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be >= 1."
`;

exports[`v4 validate options should throw an error on the "width" option with 0.1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be >= 1."
`;

exports[`v4 validate options should throw an error on the "width" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be a number (should be >= 1)."
`;

exports[`v5 validate options should throw an error on the "background" option with 9 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v5 validate options should throw an error on the "background" option with false value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v5 validate options should throw an error on the "fileLoader" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoader should be a string."
`;

exports[`v5 validate options should throw an error on the "fileLoaderOptionsGenerator" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fileLoaderOptionsGenerator should be one of these:
   string | function
   Details:
    * options.fileLoaderOptionsGenerator should be a string.
    * options.fileLoaderOptionsGenerator should be an instance of function."
`;

exports[`v5 validate options should throw an error on the "fit" option with "0" value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fit should be one of these:
   \\"cover\\" | \\"contain\\" | \\"fill\\" | \\"inside\\" | \\"outside\\""
`;

exports[`v5 validate options should throw an error on the "fit" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.fit should be one of these:
   \\"cover\\" | \\"contain\\" | \\"fill\\" | \\"inside\\" | \\"outside\\""
`;

exports[`v5 validate options should throw an error on the "format" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.format should be one of these:
   \\"jpeg\\" | \\"png\\" | \\"webp\\" | \\"avif\\" | \\"tiff\\""
`;

exports[`v5 validate options should throw an error on the "height" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be >= 1."
`;

exports[`v5 validate options should throw an error on the "height" option with 0.1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be >= 1."
`;

exports[`v5 validate options should throw an error on the "height" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.height should be a number (should be >= 1)."
`;

exports[`v5 validate options should throw an error on the "position" option with "0" value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.position should be one of these:
   \\"top\\" | \\"right top\\" | \\"right\\" | \\"right bottom\\" | \\"bottom\\" | \\"left bottom\\" | \\"left\\" | \\"left top\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\" | \\"center\\" | \\"centre\\" | \\"entropy\\" | \\"attention\\""
`;

exports[`v5 validate options should throw an error on the "position" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.position should be one of these:
   \\"top\\" | \\"right top\\" | \\"right\\" | \\"right bottom\\" | \\"bottom\\" | \\"left bottom\\" | \\"left\\" | \\"left top\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\" | \\"center\\" | \\"centre\\" | \\"entropy\\" | \\"attention\\""
`;

exports[`v5 validate options should throw an error on the "quality" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.quality should be >= 1."
`;

exports[`v5 validate options should throw an error on the "scale" option with -1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scale should be > 0."
`;

exports[`v5 validate options should throw an error on the "scale" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scale should be > 0."
`;

exports[`v5 validate options should throw an error on the "scaleUp" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.scaleUp should be a boolean."
`;

exports[`v5 validate options should throw an error on the "sharpOptions" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.sharpOptions should be an object:
   object { resize?, png?, jpeg?, webp?, tiff?, … }"
`;

exports[`v5 validate options should throw an error on the "width" option with 0 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be >= 1."
`;

exports[`v5 validate options should throw an error on the "width" option with 0.1 value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be >= 1."
`;

exports[`v5 validate options should throw an error on the "width" option with true value 1`] = `
"Invalid options object. Image Resize Loader has been initialized using an options object that does not match the API schema.
 - options.width should be a number (should be >= 1)."
`;
